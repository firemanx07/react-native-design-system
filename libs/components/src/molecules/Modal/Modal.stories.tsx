import { ComponentMeta, ComponentStory } from '@storybook/react';
import React, { useState } from 'react';

import { BaseButton } from '../../atoms';
import { StorybookScreen, StorybookSection } from '../../storybook';
import { default as ModalBase, ModalProps, ModalTypes } from './Modal';

const Story: ComponentMeta<typeof ModalBase> = {
  component: ModalBase,
  title: 'Molecules/Modal',
};
export default Story;

const noop = () => {
  // This is a mock
};

const Template: ComponentStory<typeof ModalBase> = (args: ModalProps) => {
  const [visible, setVisible] = useState(true);
  const onPress = () => setVisible(true);
  const onClose = () => setVisible(false);
  return (
    <StorybookScreen title="Modal Button:">
      <StorybookSection>
        <BaseButton onPress={onPress}>showModal</BaseButton>
        <ModalBase {...args} visible={visible} onClose={onClose} />
      </StorybookSection>
    </StorybookScreen>
  );
};

const parameters = {
  controls: {
    include: ['type', 'title', 'message', 'buttonLabel', 'OTPLabel'],
  },
};

const defaultArgs: ModalProps = {
  visible: true,
  title: 'Your request has been rejected',
  message: 'Thank you for your patience',
  type: ModalTypes.ERROR,
  buttonLabel: 'Confirm',
  OTPLabel: 'Enter your OTP code',
  onClose: noop,
};

export const Modal = Template.bind({});
Modal.parameters = parameters;
Modal.args = {
  ...defaultArgs,
};
