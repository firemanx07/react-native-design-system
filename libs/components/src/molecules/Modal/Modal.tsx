import { styled, useNamespacedTheme } from '@proxym/themes';
import { BlurView } from '@react-native-community/blur';
import React, { memo, useCallback, useMemo, useState } from 'react';
import { StyleSheet } from 'react-native';

import {
  ActionIcon,
  BaseButton,
  BaseText,
  ButtonSize,
  CloseIcon,
  ErrorIcon,
  IconButton,
  SuccessIcon,
  TextVariant,
  TextWeight,
} from '../../atoms';
import { TextInputField } from '../TextInputField';

export enum ModalTypes {
  SUCCESS = 'success',
  ERROR = 'error',
  ACTION = 'action',
  OTP = 'otp',
}
const CLOSE_ICON_SIZE = 11;
const MODAL_ICON_WIDTH = 71;
const MODAL_ICON_HEIGHT = 60;
const OTP_PLACEHOLDER = '12345';
export type ModalProps = {
  visible: boolean;
  width?: number;
  type: ModalTypes;
  title: string;
  message?: string;
  buttonLabel?: string;
  onClose: () => void;
  onSubmit?: (val?: string) => void;
  OTPLabel?: string;
  enableBackDrop?: boolean;
};

const Modal = ({
  visible,
  type,
  message = '',
  onClose,
  title,
  buttonLabel = 'Done',
  onSubmit,
  OTPLabel = 'Enter your OTP code',
  width = 340,
}: ModalProps) => {
  const { colors } = useNamespacedTheme();
  const [otpValue, setOtpValue] = useState<string>('');

  const otpTextInputWidth = useMemo<number>(() => (270 * width) / 340, [width]);
  const iconModal = useMemo<React.ReactNode | null>(() => {
    switch (type) {
      case ModalTypes.SUCCESS:
        return (
          <SuccessIcon width={MODAL_ICON_WIDTH} height={MODAL_ICON_HEIGHT} />
        );
      case ModalTypes.ERROR:
        return (
          <ErrorIcon width={MODAL_ICON_WIDTH} height={MODAL_ICON_HEIGHT} />
        );
      case ModalTypes.ACTION:
        return (
          <ActionIcon width={MODAL_ICON_WIDTH} height={MODAL_ICON_HEIGHT} />
        );
      case ModalTypes.OTP:
      default:
        return null;
    }
  }, [type]);

  const renderIcon = useCallback<
    (size: number, color: string) => React.ReactNode
  >((size, color) => <CloseIcon width={size} height={size} fill={color} />, []);
  const handleSubmit = useCallback(() => {
    onSubmit && onSubmit(type === ModalTypes.OTP ? otpValue : undefined);
    onClose();
  }, [onClose, onSubmit, otpValue, type]);

  return (
    <BaseModal
      presentationStyle="overFullScreen"
      visible={visible}
      transparent
      animationType="fade"
      onRequestClose={onClose}
    >
      <Container>
        <BlurView
          style={StyleSheet.absoluteFillObject}
          blurType="light"
          blurAmount={10}
          reducedTransparencyFallbackColor="white"
        />
        <Backdrop onPress={onClose} />
        <StyledModal width={width}>
          <StyledIconButton
            renderIcon={renderIcon}
            onPress={onClose}
            iconColor={colors.dark}
            iconSize={CLOSE_ICON_SIZE}
          />
          <ModalBody>
            {iconModal}
            <BodyContainer>
              <BaseText variant={TextVariant.L} weight={TextWeight.Medium}>
                {title}
              </BaseText>
              {type !== ModalTypes.OTP && (
                <BaseText variant={TextVariant.M}>{message}</BaseText>
              )}
            </BodyContainer>
            {type === ModalTypes.OTP && (
              <TextInputField
                value={otpValue}
                onChangeText={setOtpValue}
                label={OTPLabel}
                width={otpTextInputWidth}
                placeholder={OTP_PLACEHOLDER}
              />
            )}
            {[ModalTypes.OTP, ModalTypes.ACTION].includes(type) && (
              <BodyContainer>
                <BaseButton size={ButtonSize.Medium} onPress={handleSubmit}>
                  {buttonLabel}
                </BaseButton>
              </BodyContainer>
            )}
          </ModalBody>
        </StyledModal>
      </Container>
    </BaseModal>
  );
};

const BaseModal = styled.Modal``;
const StyledIconButton = styled(IconButton)`
  background-color: transparent;
  position: absolute;
  right: 0;
  z-index: 1;
`;
const BodyContainer = styled.View`
  align-items: center;
  justify-content: center;
  gap: ${({ theme }) => theme.ds.spacing.small}px;
`;
const Container = styled.View`
  flex: 1;
  align-items: center;
  justify-content: center;
  background-color: rgba(0, 0, 0, 0.22);
  backdrop-filter: blur(8px);
`;
const ModalBody = styled.View`
  padding-vertical: ${({ theme }) => theme.ds.spacing.huge}px;
  gap: ${({ theme }) => theme.ds.spacing.large}px;
  align-items: center;
`;
type StyledModalProp = {
  width: number;
};
const StyledModal = styled.View<StyledModalProp>`
  width: ${({ width }) => width}px;
  min-height: 238px;
  background-color: rgba(255, 255, 255, 0.93);
  shadow-color: rgba(44, 44, 44, 0.18);
  shadow-offset: 10px 35px;
  shadow-opacity: 1;
  border-radius: 7px;
  shadow-radius: 40px;
`;
const Backdrop = styled.TouchableOpacity`
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
`;

export default memo(Modal);
