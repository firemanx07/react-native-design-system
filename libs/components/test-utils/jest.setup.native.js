import { Animated } from 'react-native';

require('react-native-reanimated/lib/reanimated2/jestUtils').setUpTests();
require('react-native-gesture-handler/jestSetup.js');

console.warn = () => {
  //this is mock
};

Animated['timing'] = (value, config) => ({
  start: callback => {
    value.setValue(config.toValue);
    callback && callback({ finished: true });
  },
  stop: jest.fn(),
});
