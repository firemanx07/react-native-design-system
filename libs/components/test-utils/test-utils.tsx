import { theme, ThemeProvider } from '@celebrate-app/themes';
import { Story } from '@storybook/react';
import {
  cleanup,
  render,
  RenderOptions,
  waitFor,
} from '@testing-library/react-native';
import React, { ReactNode } from 'react';

const Providers = ({ children }: { children: ReactNode }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

export const customRender = (
  ui: React.ReactElement<any>,
  options?: RenderOptions
) => {
  return render(ui, { wrapper: Providers, ...options });
};

beforeEach(cleanup);
afterEach(cleanup);

export function testStory(displayName: string, TestComponent: Story) {
  test(`${displayName} snapshot test`, async () => {
    const comp = customRender(<TestComponent {...TestComponent.args} />);
    jest.runOnlyPendingTimers();

    await waitFor(() => {
      expect(comp).toBeTruthy();
      expect(comp.toJSON()).toMatchSnapshot();
    });
  });
}

export function testComponentSnapshot(
  TestComponent: React.ReactElement<
    any,
    string | React.JSXElementConstructor<any>
  >
) {
  test('Snapshot test', () => {
    const comp = customRender(TestComponent);
    expect(comp).toBeTruthy();
    expect(comp.toJSON()).toMatchSnapshot();
  });
}

export function testAllStories(stories: any) {
  Object.keys(stories).forEach((story) => {
    if (story !== 'default') {
      testStory(story, stories[story]);
    }
  });
}
