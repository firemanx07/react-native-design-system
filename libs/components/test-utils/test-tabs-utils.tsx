import { waitFor } from '@testing-library/react-native';

import { customRender } from './test-utils';

type renderTabsProps = {
  elem: JSX.Element;
  testIDs: string[];
};
export const renderTabs = ({ elem, testIDs }: renderTabsProps) => {
  const comp = customRender(elem);
  const tabs: ReturnType<typeof comp.getByTestId>[] = [];
  testIDs.forEach(testID => tabs.push(comp.getByTestId(testID)));
  return { comp, tabs };
};
export const testTabs = async (args: ReturnType<typeof renderTabs>) => {
  await waitFor(() => {
    expect(args.comp).toBeTruthy();
    args.tabs.forEach(tab => expect(tab).toBeDefined());
  });
};
