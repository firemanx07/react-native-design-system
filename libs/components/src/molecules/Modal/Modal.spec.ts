import { testAllStories } from '../../../test-utils';
import * as stories from './Modal.stories';

describe('Modal tests', () => {
  testAllStories(stories);
});
