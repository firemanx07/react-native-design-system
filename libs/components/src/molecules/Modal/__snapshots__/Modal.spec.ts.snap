// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal tests Modal snapshot test 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#FFFFFF",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "minHeight": 600,
        "paddingBottom": 16,
        "paddingLeft": 16,
        "paddingRight": 16,
        "paddingTop": 16,
      },
    ]
  }
>
  <Text
    fontFamily="Rubik-Bold"
    fontSize={20}
    lineHeight={26}
    style={
      [
        {
          "color": "#2C2C2C",
          "fontFamily": "Rubik-Bold",
          "fontSize": 20,
          "lineHeight": 26,
        },
        {
          "marginBottom": 24,
        },
      ]
    }
  >
    Modal Button:
  </Text>
  <View
    style={
      [
        {
          "borderColor": "#2C2C2C",
          "borderRadius": 4,
          "borderStyle": "dashed",
          "borderWidth": 1,
          "marginBottom": 24,
          "marginLeft": 24,
          "marginRight": 24,
          "marginTop": 24,
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
          "width": 370,
        },
      ]
    }
    width={370}
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#4D37D9",
          "borderRadius": 7,
          "flexDirection": "row",
          "height": 48,
          "justifyContent": "center",
          "maxWidth": "100%",
          "opacity": 1,
          "paddingHorizontal": 24,
        }
      }
    >
      <Text
        color="#FFFFFF"
        fontFamily="Rubik-Bold"
        fontSize={16}
        hasIcon={false}
        lineHeight={20}
        numberOfLines={1}
        size="Regular"
        style={
          [
            {
              "color": "#2C2C2C",
              "fontFamily": "Rubik-Bold",
              "fontSize": 16,
              "lineHeight": 20,
            },
            {
              "color": "#FFFFFF",
              "marginLeft": 0,
            },
          ]
        }
      >
        showModal
      </Text>
    </View>
    <Modal
      animationType="fade"
      hardwareAccelerated={false}
      onRequestClose={[Function]}
      presentationStyle="overFullScreen"
      style={
        [
          {},
        ]
      }
      transparent={true}
      visible={true}
    >
      <View
        style={
          [
            {
              "alignItems": "center",
              "backdropFilter": "blur(8px)",
              "backgroundColor": "rgba(0, 0, 0, 0.22)",
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
              "justifyContent": "center",
            },
          ]
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "bottom": 0,
              "left": 0,
              "opacity": 1,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        />
        <View
          style={
            [
              {
                "backgroundColor": "rgba(255, 255, 255, 0.93)",
                "borderRadius": 7,
                "minHeight": 238,
                "shadowColor": "rgba(44, 44, 44, 0.18)",
                "shadowOffset": {
                  "height": 35,
                  "width": 10,
                },
                "shadowOpacity": 1,
                "shadowRadius": 40,
                "width": 340,
              },
            ]
          }
          width={340}
        >
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "borderRadius": 7,
                "height": 34,
                "justifyContent": "center",
                "opacity": 1,
                "position": "absolute",
                "right": 0,
                "width": 34,
                "zIndex": 1,
              }
            }
          >
            <test-file-stub
              fill="#2C2C2C"
              height={11}
              width={11}
            />
          </View>
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "gap": 28,
                  "paddingVertical": 40,
                },
              ]
            }
          >
            <test-file-stub
              height={60}
              width={71}
            />
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "gap": 8,
                    "justifyContent": "center",
                  },
                ]
              }
            >
              <Text
                fontFamily="Rubik-Medium"
                fontSize={16}
                lineHeight={20}
                style={
                  [
                    {
                      "color": "#2C2C2C",
                      "fontFamily": "Rubik-Medium",
                      "fontSize": 16,
                      "lineHeight": 20,
                    },
                  ]
                }
              >
                Your request has been rejected
              </Text>
              <Text
                fontFamily="Rubik-Regular"
                fontSize={14}
                lineHeight={20}
                style={
                  [
                    {
                      "color": "#2C2C2C",
                      "fontFamily": "Rubik-Regular",
                      "fontSize": 14,
                      "lineHeight": 20,
                    },
                  ]
                }
              >
                Thank you for your patience
              </Text>
            </View>
          </View>
        </View>
      </View>
    </Modal>
  </View>
</View>
`;
